panel/io/cache.py

```diff
@@ -305,11 +308,46 @@ def compute_hash(func, hash_funcs, args, kwargs):
         _HASH_MAP[key] = hash_value
     return hash_value
 
+P = ParamSpec("P")
+R = TypeVar("R")
+T = TypeVar("T")
+
+@overload
```

Comment:
After merging this PR with 7081 the errors I see for the `panel/io/cache.py` file are

```bash
panel/io/cache.py:38: error: Need type annotation for "_HASH_MAP" (hint: "_HASH_MAP: dict[<type>, <type>] = ...")  [var-annotated]
panel/io/cache.py:40: error: Need type annotation for "_HASH_STACKS"  [var-annotated]
panel/io/cache.py:177: error: Incompatible types in assignment (expression has type "bytes", target has type "Callable[[Any], Any]")  [assignment]
panel/io/cache.py:389: error: Incompatible return value type (got "Callable[[Any], Callable[[Callable[P, R]], Callable[P, R]]]", expected "Callable[P, R@cache] | Callable[[Callable[P, R@cache]], Callable[P, R@cache]]")  [return-value]
panel/io/cache.py:462: error: Incompatible types in "await" (actual type "R", expected type "Awaitable[Any]")  [misc]
panel/io/cache.py:493: error: "_Wrapped[P, R, P, Coroutine[Any, Any, R]]" has no attribute "clear"  [attr-defined]
panel/io/cache.py:505: error: Incompatible return value type (got "_Wrapped[P, R, P, Coroutine[Any, Any, R]]", expected "Callable[P, R] | Callable[[Callable[P, R]], Callable[P, R]]")  [return-value]
```

Thus I see some issue in line 389 after the cache type annoation is introduced. I don't see issues introduced in other files.

Reply:
Maybe a relevant youtube video:  https://www.youtube.com/watch?v=_QXlbwRmqgI

Reply:
The line 389 issue I fixed by converting the `lambda` into a regular `def`, which supports annotations.

I took an initial look at fixing the others as well:

- Added an annotation for `_HASH_MAP`
- `_HASH_STACKS` does not even appear to be used anywhere, so I couldn't figure out what it was supposed to be annotated as. `_STACK` is likewise not used anywhere. Can these be deleted?
- The line 177 error sure looks like a bug? How can a hash function be a bytes constant (see snippet below)? Those are not callable and `hash_func` is used as a callable.
```
for name in _FFI_TYPE_NAMES:
    _hash_funcs[name] = b'0'
```
- The line 493 error I ignored for now, as it is assigned an arbitrary attribute `clear` on a function. We might be able to define a more precise return type that annotates the `clear` as well but that is probably overkill.
- The remaining `async` / `await` errors I partially fixed by adding an `Awaitable` variant, however I had to add a `type: ignore` for one of them because `iscoroutinefunction` does not implement a `TypeGuard`, which it would need to to avoid the ignore. (note that `inspect.iscoroutinefunction` does implement this in `typeshed`).